base/asyncmap.jl

```diff
@@ -3,161 +3,365 @@
+                end
+            end
+        catch e
+            close(chnl)
+            retval = e
         end
-        close(itr.task_chnl)
-    catch
+        retval
     end
-    nothing
+    push!(worker_tasks, t)
 end
 
-# Open a @sync block and initialise iterator state.
-function start(itr::AsyncCollector)
-    sync_begin()
-
-    state = AsyncCollectorState(start(itr.enumerator),  0, Condition(), false, false, 0)
+# Special handling for some types.
+function asyncmap(f, s::AbstractString; kwargs...)
+    s2=Array(Char, length(s))
+    asyncmap!(f, s2, s; kwargs...)
+    return convert(String, s2)
+end
 
-    for _ in 1:itr.max_tasks()
-        start_collector_task(itr, state)
-        state.nfree += 1
+# map on a single BitArray returns a BitArray if the mapping function is boolean.
+function asyncmap(f, b::BitArray; kwargs...)
+    b2 = async_usemap(f, b; kwargs...)
+    if eltype(b2) == Bool
+        return BitArray(b2)
     end
-    state
+    return b2
 end
 
-# Close @sync block when iterator is done.
-function done(itr::AsyncCollector, state::AsyncCollectorState)
-    if state.in_error
-        @assert !isopen(itr.task_chnl)  # Channel should have been cleared and closed in the async task.
+# TODO: Optimize for sparse arrays
+# For now process as regular arrays and convert back
+function asyncmap(f, s::AbstractSparseArray...; kwargs...)
+    sa = map(Array, s)
+    return sparse(asyncmap(f, sa...; kwargs...))
+end
 
-        sync_end()
+type AsyncCollector
+    f
+    results
+    enumerator::Enumerate
+    ntasks
+    batch_size
+    nt_check::Bool     # check number of tasks on every iteration
 
-        # state.in_error is only being set in the @async block (and an error thrown),
-        # which in turn should have been caught and thrown by the sync_end() call above.
-        # Control should not come here.
-        @assert false "Error should have been captured and thrown previously."
-    end
+    AsyncCollector(f, r, en::Enumerate, ntasks, batch_size) = new(f, r, en, ntasks, batch_size, isa(ntasks, Function))
+end
+
+"""
+    AsyncCollector(f, results, c...; ntasks=0, batch_size=nothing) -> iterator
+
+Returns an iterator which applies `f` to each element of `c` asynchronously
+and collects output into results.
 
-    if !state.done && done(itr.enumerator, state.enum_state)
-        state.done = true
-        close(itr.task_chnl)
-        sync_end()
+Keyword args `ntasks` and `batch_size` have the same behavior as in
+[`asyncmap()`](@ref). If `batch_size` is specified, `f` must
+be a function which operates on an array of argument tuples.
+
+!!! note
+    `next(::AsyncCollector, state) -> (nothing, state)`. A successful return
+    from `next` indicates that the next element from the input collection is
+    being processed asynchronously. It blocks till a free worker task becomes
+    available.
+
+!!! note
+    `for _ in AsyncCollector(f, results, c...) end` is equivalent to
+    `map!(f, results, c...)`.
```

Comment:
Both. The former in terms of returned data. The latter in terms of asynchronous execution.